gobblin-hive-registration/src/main/java/gobblin/hive/avro/HiveAvroSerDeManager.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.hive.avro;
+
+import java.io.IOException;
+
+import org.apache.avro.Schema;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Preconditions;
+
+import gobblin.annotation.Alpha;
+import gobblin.configuration.State;
+import gobblin.hive.HiveRegistrationUnit;
+import gobblin.hive.HiveSerDeManager;
+import gobblin.hive.HiveSerDeWrapper;
+import gobblin.util.AvroUtils;
+import gobblin.util.HadoopUtils;
+
+
+/**
+ * A {@link HiveSerDeManager} for registering Avro tables and partitions.
+ *
+ * @author ziliu
+ */
+@Alpha
+public class HiveAvroSerDeManager extends HiveSerDeManager {
+
+  public static final String SCHEMA_LITERAL = "avro.schema.literal";
+  public static final String SCHEMA_URL = "avro.schema.url";
+  public static final String USE_SCHEMA_FILE = "use.schema.file";
+  public static final boolean DEFAULT_USE_SCHEMA_FILE = false;
+  public static final String SCHEMA_FILE_NAME = "schema.file.name";
+  public static final String DEFAULT_SCHEMA_FILE_NAME = "_schema.avsc";
+  public static final String SCHEMA_LITERAL_LENGTH_LIMIT = "schema.literal.length.limit";
+  public static final int DEFAULT_SCHEMA_LITERAL_LENGTH_LIMIT = 4000;
+
+  protected final FileSystem fs;
+  protected final boolean useSchemaFile;
+  protected final String schemaFileName;
+  protected final int schemaLiteralLengthLimit;
+  protected final HiveSerDeWrapper serDeWrapper = HiveSerDeWrapper.get("AVRO");
+
+  public HiveAvroSerDeManager(State props) throws IOException {
+    super(props);
+    this.fs = FileSystem.get(HadoopUtils.getConfFromState(props));
+    this.useSchemaFile = props.getPropAsBoolean(USE_SCHEMA_FILE, DEFAULT_USE_SCHEMA_FILE);
+    this.schemaFileName = props.getProp(SCHEMA_FILE_NAME, DEFAULT_SCHEMA_FILE_NAME);
+    this.schemaLiteralLengthLimit =
+        props.getPropAsInt(SCHEMA_LITERAL_LENGTH_LIMIT, DEFAULT_SCHEMA_LITERAL_LENGTH_LIMIT);
+  }
+
+  /**
+   * Add an Avro {@link Schema} to the given {@link HiveRegistrationUnit}. 
+   *
+   *  <p>
+   *    If {@link #USE_SCHEMA_FILE} is true, the schema will be added via {@link #SCHEMA_URL} pointing to
+   *    the schema file named {@link #SCHEMA_FILE_NAME}.
+   *  </p>
+   *
+   *  <p>
+   *    If {@link #USE_SCHEMA_FILE} is false, the schema will be obtained by {@link #getDirectorySchema(Path)}. 
+   *    If the length of the schema is less than {@link #SCHEMA_LITERAL_LENGTH_LIMIT}, it will be added via 
+   *    {@link #SCHEMA_LITERAL}. Otherwise, the schema will be written to {@link #SCHEMA_FILE_NAME} and added
+   *    via {@link #SCHEMA_URL}.
+   *  </p>
+   */
+  @Override
+  public void addSerDeProperties(Path path, HiveRegistrationUnit hiveUnit) throws IOException {
```

Comment:
Yes we can make it a field in `HiveRegistrationUnit`. But I'd like to keep the `HiveSerDeManager` name since it is really a manager class, e.g., it contains logic to fetch and keep track of the schemas; it is also responsible for determining whether the table schema should be updated when adding a new partition.